larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+        throw StatusCodeException( STATUS_CODE_INVALID_PARAMETER );
+    }
+    else if (std::fabs(eigenValues.GetY()) < std::numeric_limits<float>::epsilon())
+    {
+        return 0.f;
+    }
+
+    return std::atan(std::sqrt(eigenValues.GetY()) * sinTheta / std::sqrt(eigenValues.GetX()));
+}
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode OpeningAngleFeatureTool::ReadSettings(const TiXmlHandle /*xmlHandle*/)
+{
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+PCAFeatureTool::PCAFeatureTool() :
+    m_ratioVariables(true)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void PCAFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
+{
+     if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+        // Need the 3D cluster and hits to calculate PCA components
+        ClusterList threeDClusterList;
+        LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
+
+        if (threeDClusterList.empty())
+            return;
+
+        CaloHitList threeDCaloHitList; // Might be able to get through LArPfoHelper instead
+        const Cluster *const pThreeDCluster(threeDClusterList.front());
+        pThreeDCluster->GetOrderedCaloHitList().FillCaloHitList(threeDCaloHitList);
+
+        if (threeDCaloHitList.empty())
+            return;
+
+        // Run the PCA analysis
+        CartesianVector centroid(0.f, 0.f, 0.f);
+        LArPcaHelper::EigenVectors eigenVecs;
+        LArPcaHelper::EigenValues eigenValues(0.f, 0.f, 0.f);
+        LArPcaHelper::RunPca(threeDCaloHitList, centroid, eigenValues, eigenVecs);
+        const float principalEigenvalue(eigenValues.GetX()), secondaryEigenvalue(eigenValues.GetY()), tertiaryEigenvalue(eigenValues.GetZ());
+
+        if (m_ratioVariables)
+        {
+            float pca1(-1.f), pca2(-1.f);
+            if(principalEigenvalue > std::numeric_limits<float>::epsilon())
+            {
+                pca1 = secondaryEigenvalue/principalEigenvalue;
+                pca2 = tertiaryEigenvalue/principalEigenvalue;
+            }
+            featureVector.push_back(pca1);
+            featureVector.push_back(pca2);
+        }
+        else
```

Comment:
Different numbers of features with, without ratio variables?

Reply:
different indeed, but should be irrelevant if the ratioVariables off option is dropped